### ISSUE TITLE
Question, Where is the controller? Where is the domain url mapping configured?

### ISSUE THREAD

**Issue by kabasakalis** [cristinanegrean/wanderlust-open-travel-api#4]

**Comment by cristinanegrean** [comment]
Apologies for the late reaction (been quite busy), so I hope the response arrives still in time to be useful!

Good observation: there is no controller. The idea the repo tries to prove is "writing less code" to expose a domain model as hypermedia links for simple CRUD(Create/Read/Update/Delete) commands with Spring Data REST. A good example which shows explicitly the parallel between using `@RestController` versus just `@RepositoryRestResource` can be found [here](https://blog.trifork.com/2016/11/23/writing-less-code/)

I do see the need of writing a controller when:
- the hypermedia link exposes data that isn't one on one representation of your domain, example [here](https://github.com/cristinanegrean/spring-cloud-stream-kafka/blob/master/src/main/java/cristina/tech/fancydress/store/controller/TrendingRestController.java)
- combining HAL(Hypertext Application Language) with HATEOAS (Hypermedia as the Engine of Application State). In this repo only HAL is exemplified: representing resources, that will contain hypermedia information (links to related resources). If you want a very advanced example of HATEOAS, I warmly recommend this [one](https://github.com/kbastani/event-stream-processing-microservices)

The domain url mapping is configured [here](https://github.com/cristinanegrean/wanderlust-open-travel-api/blob/master/src/main/resources/application.properties). Look for `spring.data.rest.base-path` key value. Then for each repository, you can provide an explicit path, i.e: `@RepositoryRestResource(collectionResourceRel = "destinations", path = "destinations")`, but it is not mandatory! By default, Spring Data REST will expose your CrudRepository using the name of the domain class (in this case Destination), lower-cased, and apply the [Evo Inflector](https://github.com/atteo/evo-inflector) to pluralize the word.

**Comment by kabasakalis** [comment]
Thank you for your detailed explanation. I am quite new to Spring and I was searching Github repos for examples and ideas to help me build [springify = spring + spotify ](https://github.com/kabasakalis/springify), my first Spring HATEOAS API demo, which is almost finished now, only swagger docs missing.  Since then I learned  how `data-rest` automates everything, very convenient. However, I felt I was  missing full control. For example, maybe I want to do some custom stuff in my controllers, specially related on how I retrieve data from the database. Or maybe I want to include some custom links in the response. Or I want to throw custom NotFound exceptions that return rich information about the failed request. Or completely separate the model from the marshalled response with `ResourceSupport` extensions (and avoid littering the model with `@JsonIgnore` annotations or methods that exist solely for the purpose of returning extra data in addition to model properties). Kind of things you also pointed out. Your repo  perfectly showcase the fact that `data-rest` builds a full API if you are OK with the defaults, very impressive. And I learned about the custom validations, didn't use that much in my demo. I bookmarked the repositories you suggested. Again, thank you for your time, and thanks for inspiration.
